python-packages/aws-event-stream/aws_event_stream/aio/__init__.py

```diff
@@ -0,0 +1,247 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+import asyncio
+from collections.abc import Callable
+from typing import Self
+
+from smithy_core.aio.interfaces import AsyncByteStream, AsyncWriter
+from smithy_core.codecs import Codec
+from smithy_core.deserializers import DeserializeableShape, ShapeDeserializer
+from smithy_core.serializers import SerializeableShape
+from smithy_event_stream.aio.interfaces import (
+    AsyncEventReceiver,
+    DuplexEventStream,
+    InputEventStream,
+    OutputEventStream,
+)
+
+from .._private.deserializers import AWSAsyncEventReceiver as _AWSEventReceiver
+from .._private.serializers import AWSAsyncEventPublisher as _AWSEventPublisher
+from .._private.serializers import Signer
+from ..exceptions import MissingInitialResponse
+
+
+class AWSDuplexEventStream[
+    I: SerializeableShape, O: DeserializeableShape, R: DeserializeableShape
+](DuplexEventStream[I, O, R]):
+    """A duplex event stream using the application/vnd.amazon.eventstream format."""
+
+    def __init__(
+        self,
+        payload_codec: Codec,
+        async_writer: AsyncWriter,
+        deserializer: Callable[[ShapeDeserializer], O],
+        async_reader: AsyncByteStream | None = None,
+        initial_response: R | None = None,
+        deserializeable_response: type[R] | None = None,
+        signer: Signer | None = None,
+        is_client_mode: bool = True,
+    ) -> None:
+        """Construct an AWSDuplexEventStream.
+
+        :param payload_codec: The codec to encode the event payload with.
+        :param async_writer: The writer to write event bytes to.
+        :param deserializer: A callable to deserialize events with. This should be the
+            union's deserialize method.
+        :param async_reader: The reader to read event bytes from, if available. If not
+            immediately available, output will be blocked on it becoming available.
+        :param initial_response: The deserialized operation response, if available. If
+            not immediately available, output will be blocked on it becoming available.
+        :param deserializeable_response: The deserializeable response class. Setting
+            this indicates that the initial response is sent over the event stream. The
+            deserialize method of this class will be used to deserialize it upon
+            calling ``await_output``.
+        :param signer: An optional callable to sign events with prior to them being
+            encoded.
+        :param is_client_mode: Whether the stream is being constructed for a client or
+            serfer implementation.
+        """
+        self.input_stream = _AWSEventPublisher(
+            payload_codec=payload_codec,
+            async_writer=async_writer,
+            signer=signer,
+            is_client_mode=is_client_mode,
+        )
+
+        self._deserialize = deserializer
```

Comment:
Did we mean to call this `deserializer` in some APIs but `deserialize` in others?